### PR TITLE
feat: support hash and hash_kv

### PR DESCRIPTION
**Pull request by themanforfree**
Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)
support hash and hash_kv rpc in maintenance server
* what changes does this pull request make?
add hash and hash_kv implementation
* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)
no

### CODE REVIEW

File: xline/src/storage/kv_store.rs [hunk 1]
```diff
@@ -360,6 +360,43 @@ where
...
+            return Err(tonic::Status::out_of_range(
+                "etcdserver: mvcc: required revision has been compacted",
+            ));
+        }
+        if rev > 0 && rev > current_rev {
+            return Err(tonic::Status::out_of_range(
+                "etcdserver: mvcc: required revision is a future revision",
+            ));
+        }
+        if rev == 0 {
```
Comment (bsbds) [inline xline/src/storage/kv_store.rs:377]: We should set the rev to current rev if it's less or equal to zero.

File: xline/src/storage/kv_store.rs [hunk 2]
```diff
@@ -360,6 +360,43 @@ where
...
+        let mut hasher = crc32fast::Hasher::new();
+        hasher.update(KV_TABLE.as_bytes());
+        let kv_pairs = self.db.get_all(KV_TABLE)?;
+        for (k, v) in kv_pairs {
+            let kr = Revision::decode(&k);
+            let upper = Revision::new(rev.overflow_add(1), 0);
+            if upper <= kr {
+                continue;
+            }
+            let lower = Revision::new(compact_rev.overflow_add(1), 0);
```
Comment (bsbds) [inline xline/src/storage/kv_store.rs:390]: The compact revision does not needs to be incremented, compact rev refers to the lowest valid revision.

File: xline/src/storage/kv_store.rs [hunk 3]
```diff
@@ -360,6 +360,43 @@ where
...
+            if lower > kr && !keep.is_empty() && !keep.contains(&kr) {
```
Comment (bsbds) [inline xline/src/storage/kv_store.rs:391]: The `is_empty` check is redundant as `contains` implies it.

File: xline/src/storage/db.rs [hunk 4]
```diff
@@ -203,6 +203,21 @@ impl StorageApi for DB {
             .map_err(|e| ExecuteError::DbError(format!("Failed to flush ops, error: {e}")))?;
         Ok(revs)
     }
+
+    fn hash(&self) -> Result<u32, ExecuteError> {
+        let mut hasher = crc32fast::Hasher::new();
+        for table in XLINE_TABLES {
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/storage/db.rs:209]: Just for test, Same as Etcd

File: xline/src/storage/kv_store.rs (hunk 2 above)
Comment (themanforfree) [reply xline/src/storage/kv_store.rs:390]: Yes, so when `kr == compact_rev`, `lower` will be greater than `kr`, `kr` equal to `compact_rev` will be used to calculate

**Review by bsbds — DISMISSED** [review]

File: xline/src/storage/kv_store.rs [hunk 5]
```diff
@@ -479,6 +479,43 @@ where
...
+        }
+        if rev > 0 && rev > current_rev {
+            return Err(tonic::Status::out_of_range(
+                "etcdserver: mvcc: required revision is a future revision",
+            ));
+        }
+        if rev <= 0 {
+            rev = current_rev;
+        }
+        let keep = self.inner.index.keep(rev);
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/storage/kv_store.rs:499]: The logic here overlaps with `RevisionCheck`, but it's more complicated, and we need to continue using `compact_rev` and `current_rev` after check revision, so I think it is not suitable to use `RevisionCheck` here.

File: xline/src/storage/kv_store.rs [hunk 6]
```diff
@@ -479,6 +479,43 @@ where
...
+        let mut hasher = crc32fast::Hasher::new();
+        hasher.update(KV_TABLE.as_bytes());
+        let kv_pairs = self.inner.db.get_all(KV_TABLE)?;
+        for (k, v) in kv_pairs {
+            let kr = Revision::decode(&k);
+            let upper = Revision::new(rev.overflow_add(1), 0);
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/storage/kv_store.rs]: Moved

File: xline/src/storage/kv_store.rs [hunk 7]
```diff
@@ -479,6 +479,43 @@ where
...
+            if upper <= kr {
+                continue;
+            }
+            let lower = Revision::new(compact_rev, 0);
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/storage/kv_store.rs]: Moved

**Review by iGxnon — DISMISSED** [review]

File: xline/src/storage/kv_store.rs [hunk 8]
```diff
@@ -479,6 +479,43 @@ where
         _ = self.inner.db.flush_ops(ops)?;
         Ok(())
     }
+
+    /// Calculate hash of kv storage
+    pub(crate) fn hash_kv(&self, mut rev: i64) -> Result<(u32, i64, i64), tonic::Status> {
+        let (compact_rev, current_rev) = (self.compacted_revision(), self.revision());
+        if rev > 0 && rev < compact_rev {
+            return Err(tonic::Status::out_of_range(
+                "etcdserver: mvcc: required revision has been compacted",
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/storage/kv_store.rs]: FIxed

File: xline/src/storage/kv_store.rs [hunk 9]
```diff
@@ -479,6 +479,43 @@ where
...
+            ));
+        }
+        if rev > 0 && rev > current_rev {
+            return Err(tonic::Status::out_of_range(
+                "etcdserver: mvcc: required revision is a future revision",
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/storage/kv_store.rs]: Same as above

**Review by Phoenix500526 — APPROVED** [review]

**Review by bsbds — APPROVED** [review]

**Review by Phoenix500526 — APPROVED** [review]
LGTM

**Pull request merged**